{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "78765a48",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "# NCBI Genome/RefSeq ingestion script\n",
    "\n",
    "This script reads the listing for the:\n",
    "\n",
    "- NCBI genome/refseq via the URL https://ftp.ncbi.nih.gov/genomes/refseq/assembly_summary_refseq.txt \n",
    "- NCBI genome/genbank via the URL https://ftp.ncbi.nih.gov/genomes/genbank/assembly_summary_genbank.txt\n",
    "\n",
    "and attempts to map the columns to Darwin Core (DwC), using any of the DwC extenesions that might apply. More info in [this issue](https://github.com/AtlasOfLivingAustralia/arga-issues/issues/8).\n",
    "\n",
    "### TODO\n",
    "- Moved into https://github.com/ARGA-Genomes/arga-data/issues/2#issuecomment-1185139354"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7b611d7c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d6068f75ae4f4cc6ad64f0e2a3895cb0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "RadioButtons(description='DB option:', layout=Layout(width='max-content'), options=('refseq', 'genbank'), valu…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import ipywidgets as widgets\n",
    "\n",
    "dbName = widgets.RadioButtons(\n",
    "    options=['refseq', 'genbank'],\n",
    "    value='refseq', # Defaults to 'refseq'\n",
    "    layout={'width': 'max-content'}, # If the items' names are long\n",
    "    description='DB option:',\n",
    "    disabled=False\n",
    ")\n",
    "\n",
    "display(dbName)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Selected DB = genbank\n"
     ]
    }
   ],
   "source": [
    "print(f\"Selected DB = {dbName.value}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d23b6114",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "genome_directory = dbName.value\n",
    "# genome_directory = \"genbank\"\n",
    "df = pd.read_table(f\"https://ftp.ncbi.nih.gov/genomes/{genome_directory}/assembly_summary_{genome_directory}.txt\", header=1, dtype='object', parse_dates=['seq_rel_date'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24589834",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Look at the data types detected by pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "eb71cbdc",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 272351 entries, 0 to 272350\n",
      "Data columns (total 23 columns):\n",
      " #   Column                     Non-Null Count   Dtype         \n",
      "---  ------                     --------------   -----         \n",
      " 0   # assembly_accession       272351 non-null  object        \n",
      " 1   bioproject                 272299 non-null  object        \n",
      " 2   biosample                  260835 non-null  object        \n",
      " 3   wgs_master                 227449 non-null  object        \n",
      " 4   refseq_category            272351 non-null  object        \n",
      " 5   taxid                      272351 non-null  object        \n",
      " 6   species_taxid              272351 non-null  object        \n",
      " 7   organism_name              272351 non-null  object        \n",
      " 8   infraspecific_name         253148 non-null  object        \n",
      " 9   isolate                    18013 non-null   object        \n",
      " 10  version_status             272351 non-null  object        \n",
      " 11  assembly_level             272351 non-null  object        \n",
      " 12  release_type               272351 non-null  object        \n",
      " 13  genome_rep                 272351 non-null  object        \n",
      " 14  seq_rel_date               272351 non-null  datetime64[ns]\n",
      " 15  asm_name                   272155 non-null  object        \n",
      " 16  submitter                  272299 non-null  object        \n",
      " 17  gbrs_paired_asm            272351 non-null  object        \n",
      " 18  paired_asm_comp            272351 non-null  object        \n",
      " 19  ftp_path                   272351 non-null  object        \n",
      " 20  excluded_from_refseq       509 non-null     object        \n",
      " 21  relation_to_type_material  27668 non-null   object        \n",
      " 22  asm_not_live_date          272351 non-null  object        \n",
      "dtypes: datetime64[ns](1), object(22)\n",
      "memory usage: 47.8+ MB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "88f47c8a",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># assembly_accession</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>wgs_master</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>taxid</th>\n",
       "      <th>species_taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>infraspecific_name</th>\n",
       "      <th>isolate</th>\n",
       "      <th>...</th>\n",
       "      <th>genome_rep</th>\n",
       "      <th>seq_rel_date</th>\n",
       "      <th>asm_name</th>\n",
       "      <th>submitter</th>\n",
       "      <th>gbrs_paired_asm</th>\n",
       "      <th>paired_asm_comp</th>\n",
       "      <th>ftp_path</th>\n",
       "      <th>excluded_from_refseq</th>\n",
       "      <th>relation_to_type_material</th>\n",
       "      <th>asm_not_live_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>272341</th>\n",
       "      <td>GCF_943908275.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA110068406</td>\n",
       "      <td>CALTSQ000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>146919</td>\n",
       "      <td>146919</td>\n",
       "      <td>Salinibacter ruber</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ES_FV_43</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2022-06-14</td>\n",
       "      <td>ASM94390827v1</td>\n",
       "      <td>IMEDEA</td>\n",
       "      <td>GCA_943908275.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272342</th>\n",
       "      <td>GCF_943908285.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA110068400</td>\n",
       "      <td>CALTSP000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>146919</td>\n",
       "      <td>146919</td>\n",
       "      <td>Salinibacter ruber</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ES_FV_35</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2022-06-14</td>\n",
       "      <td>ASM94390828v1</td>\n",
       "      <td>IMEDEA</td>\n",
       "      <td>GCA_943908285.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272343</th>\n",
       "      <td>GCF_943908305.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA110068429</td>\n",
       "      <td>CALTSU000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>146919</td>\n",
       "      <td>146919</td>\n",
       "      <td>Salinibacter ruber</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ES_FV_73</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2022-06-14</td>\n",
       "      <td>ASM94390830v1</td>\n",
       "      <td>IMEDEA</td>\n",
       "      <td>GCA_943908305.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272344</th>\n",
       "      <td>GCF_944470325.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA110148482</td>\n",
       "      <td>CALYCM000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CRE_166F4</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2022-07-13</td>\n",
       "      <td>CRE_166F4</td>\n",
       "      <td>insitut pasteur</td>\n",
       "      <td>GCA_944470325.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272345</th>\n",
       "      <td>GCF_944611425.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA110180405</td>\n",
       "      <td>CALYLZ000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>562</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>strain=PBIO3502</td>\n",
       "      <td>27-ESBL-EC</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2022-07-13</td>\n",
       "      <td>PBIO3502</td>\n",
       "      <td>University of Greifswald, Institute of Pharmacy</td>\n",
       "      <td>GCA_944611425.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272346</th>\n",
       "      <td>GCF_944739345.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA110191398</td>\n",
       "      <td>CALYGX000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>2954240</td>\n",
       "      <td>2954240</td>\n",
       "      <td>Fusobacterium sp. Marseille-Q7035</td>\n",
       "      <td>strain=Marseille-Q7035</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2022-07-13</td>\n",
       "      <td>Q7035</td>\n",
       "      <td>IHU - Mediterranee Infection</td>\n",
       "      <td>GCA_944739345.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272347</th>\n",
       "      <td>GCF_944801835.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA110212502</td>\n",
       "      <td>CALYLN000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>2911422</td>\n",
       "      <td>2911422</td>\n",
       "      <td>Bartonella sp. B1099</td>\n",
       "      <td>strain=Marseille-B1099</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2022-07-13</td>\n",
       "      <td>B1099</td>\n",
       "      <td>IHU - Mediterranee Infection</td>\n",
       "      <td>GCA_944801835.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272348</th>\n",
       "      <td>GCF_944896145.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA110212597</td>\n",
       "      <td>CALYLP000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>2911421</td>\n",
       "      <td>2911421</td>\n",
       "      <td>Bartonella sp. B1098</td>\n",
       "      <td>strain=Marseille-B1098</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2022-07-13</td>\n",
       "      <td>B1098</td>\n",
       "      <td>IHU - Mediterranee Infection</td>\n",
       "      <td>GCA_944896145.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272349</th>\n",
       "      <td>GCF_944989955.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA110191399</td>\n",
       "      <td>CALYLV000000000.1</td>\n",
       "      <td>na</td>\n",
       "      <td>1496</td>\n",
       "      <td>1496</td>\n",
       "      <td>Clostridioides difficile</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Clostridioides difficile</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2022-07-13</td>\n",
       "      <td>JMR5_assembly</td>\n",
       "      <td>Leiden University Medical Center</td>\n",
       "      <td>GCA_944989955.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272350</th>\n",
       "      <td>GCF_945605565.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA14509629</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>130049</td>\n",
       "      <td>130049</td>\n",
       "      <td>Paenibacillus dendritiformis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ena-SAMPLE-TAB-03-06-2022-11:19:07:959-6672</td>\n",
       "      <td>...</td>\n",
       "      <td>Full</td>\n",
       "      <td>2022-07-13</td>\n",
       "      <td>LMG_21716</td>\n",
       "      <td>Agroscope</td>\n",
       "      <td>GCA_945605565.1</td>\n",
       "      <td>identical</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       # assembly_accession   bioproject       biosample         wgs_master  \\\n",
       "272341      GCF_943908275.1  PRJNA224116  SAMEA110068406  CALTSQ000000000.1   \n",
       "272342      GCF_943908285.1  PRJNA224116  SAMEA110068400  CALTSP000000000.1   \n",
       "272343      GCF_943908305.1  PRJNA224116  SAMEA110068429  CALTSU000000000.1   \n",
       "272344      GCF_944470325.1  PRJNA224116  SAMEA110148482  CALYCM000000000.1   \n",
       "272345      GCF_944611425.1  PRJNA224116  SAMEA110180405  CALYLZ000000000.1   \n",
       "272346      GCF_944739345.1  PRJNA224116  SAMEA110191398  CALYGX000000000.1   \n",
       "272347      GCF_944801835.1  PRJNA224116  SAMEA110212502  CALYLN000000000.1   \n",
       "272348      GCF_944896145.1  PRJNA224116  SAMEA110212597  CALYLP000000000.1   \n",
       "272349      GCF_944989955.1  PRJNA224116  SAMEA110191399  CALYLV000000000.1   \n",
       "272350      GCF_945605565.1  PRJNA224116   SAMEA14509629                NaN   \n",
       "\n",
       "       refseq_category    taxid species_taxid  \\\n",
       "272341              na   146919        146919   \n",
       "272342              na   146919        146919   \n",
       "272343              na   146919        146919   \n",
       "272344              na      562           562   \n",
       "272345              na      562           562   \n",
       "272346              na  2954240       2954240   \n",
       "272347              na  2911422       2911422   \n",
       "272348              na  2911421       2911421   \n",
       "272349              na     1496          1496   \n",
       "272350              na   130049        130049   \n",
       "\n",
       "                            organism_name      infraspecific_name  \\\n",
       "272341                 Salinibacter ruber                     NaN   \n",
       "272342                 Salinibacter ruber                     NaN   \n",
       "272343                 Salinibacter ruber                     NaN   \n",
       "272344                   Escherichia coli                     NaN   \n",
       "272345                   Escherichia coli         strain=PBIO3502   \n",
       "272346  Fusobacterium sp. Marseille-Q7035  strain=Marseille-Q7035   \n",
       "272347               Bartonella sp. B1099  strain=Marseille-B1099   \n",
       "272348               Bartonella sp. B1098  strain=Marseille-B1098   \n",
       "272349           Clostridioides difficile                     NaN   \n",
       "272350       Paenibacillus dendritiformis                     NaN   \n",
       "\n",
       "                                            isolate  ... genome_rep  \\\n",
       "272341                                     ES_FV_43  ...       Full   \n",
       "272342                                     ES_FV_35  ...       Full   \n",
       "272343                                     ES_FV_73  ...       Full   \n",
       "272344                                    CRE_166F4  ...       Full   \n",
       "272345                                   27-ESBL-EC  ...       Full   \n",
       "272346                                          NaN  ...       Full   \n",
       "272347                                          NaN  ...       Full   \n",
       "272348                                          NaN  ...       Full   \n",
       "272349                     Clostridioides difficile  ...       Full   \n",
       "272350  ena-SAMPLE-TAB-03-06-2022-11:19:07:959-6672  ...       Full   \n",
       "\n",
       "       seq_rel_date       asm_name  \\\n",
       "272341   2022-06-14  ASM94390827v1   \n",
       "272342   2022-06-14  ASM94390828v1   \n",
       "272343   2022-06-14  ASM94390830v1   \n",
       "272344   2022-07-13      CRE_166F4   \n",
       "272345   2022-07-13       PBIO3502   \n",
       "272346   2022-07-13          Q7035   \n",
       "272347   2022-07-13          B1099   \n",
       "272348   2022-07-13          B1098   \n",
       "272349   2022-07-13  JMR5_assembly   \n",
       "272350   2022-07-13      LMG_21716   \n",
       "\n",
       "                                              submitter  gbrs_paired_asm  \\\n",
       "272341                                           IMEDEA  GCA_943908275.1   \n",
       "272342                                           IMEDEA  GCA_943908285.1   \n",
       "272343                                           IMEDEA  GCA_943908305.1   \n",
       "272344                                  insitut pasteur  GCA_944470325.1   \n",
       "272345  University of Greifswald, Institute of Pharmacy  GCA_944611425.1   \n",
       "272346                     IHU - Mediterranee Infection  GCA_944739345.1   \n",
       "272347                     IHU - Mediterranee Infection  GCA_944801835.1   \n",
       "272348                     IHU - Mediterranee Infection  GCA_944896145.1   \n",
       "272349                 Leiden University Medical Center  GCA_944989955.1   \n",
       "272350                                        Agroscope  GCA_945605565.1   \n",
       "\n",
       "       paired_asm_comp                                           ftp_path  \\\n",
       "272341       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "272342       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "272343       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "272344       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "272345       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "272346       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "272347       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "272348       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "272349       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "272350       identical  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/9...   \n",
       "\n",
       "       excluded_from_refseq relation_to_type_material asm_not_live_date  \n",
       "272341                  NaN                       NaN                na  \n",
       "272342                  NaN                       NaN                na  \n",
       "272343                  NaN                       NaN                na  \n",
       "272344                  NaN                       NaN                na  \n",
       "272345                  NaN                       NaN                na  \n",
       "272346                  NaN                       NaN                na  \n",
       "272347                  NaN                       NaN                na  \n",
       "272348                  NaN                       NaN                na  \n",
       "272349                  NaN                       NaN                na  \n",
       "272350                  NaN                       NaN                na  \n",
       "\n",
       "[10 rows x 23 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.tail(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6cb816dc",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## Field mappings\n",
    "\n",
    "NCBI fields described at https://ftp.ncbi.nih.gov/genomes/README_assembly_summary.txt\n",
    "\n",
    "| Column | field name | field description |\n",
    "|--------| ---- | ---- |\n",
    "| 1      | \"assembly_accession\" | Assembly accession: the assembly accession.version reported in this field is a unique identifier for the set of sequences in this particular version of the genome assembly. |\n",
    "| 2      | \"bioproject\" | BioProject: accession for the BioProject which produced the sequences in the genome assembly. A BioProject is a collection of biological data related to a single initiative, originating from a single organization or from a consortium. A BioProject record provides users a single place to find links to the diverse data types generated for that project. The record can be retrieved from the NCBI BioProject resource: https://www.ncbi.nlm.nih.gov/bioproject/ |\n",
    "| 3      | \"biosample\" | BioSample: accession for the BioSample from which the sequences in the genome assembly were obtained. A BioSample record contains a description of the biological source material used in experimental assays. The record can be retrieved from the NCBI BioSample resource: https://www.ncbi.nlm.nih.gov/biosample/ |\n",
    "| 4      | \"wgs_master\" | WGS-master: the GenBank Nucleotide accession and version for the master record of the Whole Genome Shotgun (WGS) project for the genome assembly. The master record can be retrieved from the NCBI Nucleotide resource: https://www.ncbi.nlm.nih.gov/nuccore Genome assemblies that are complete genomes, and those that are clone-based, do not have WGS-master records in which case this field will be empty. |\n",
    "| 5      | \"refseq_category\" | RefSeq Category: whether the assembly is a reference or representative genome in the NCBI Reference Sequence (RefSeq) project classification. Values: **reference genome** - a manually selected high quality genome assembly that NCBI and the community have identified as being important as a standard against which other data are compared; **representative genome** - a genome computationally or manually selected as a representative from among the best genomes available for a species or clade that does not have a designated reference genome; **na** - no RefSeq category assigned to this assembly Prokaryotes may have more than one reference or representative genome per species. For more information see: https://www.ncbi.nlm.nih.gov/refseq/about/prokaryotes/#referencegenome Eukaryotes have no more than one reference or representative genome per species. If there are no assemblies in RefSeq for a particular eukaryotic species, then the GenBank assembly that RefSeq would select as the best available for that species will be designated as the representative genome. Viruses may have one or more reference genomes per species. The representative genome designation is not applied to viruses. |\n",
    "| 6      | \"taxid\" | Taxonomy ID: the NCBI taxonomy identifier for the organism from which the genome assembly was derived. The NCBI Taxonomy Database is a curated classification and nomenclature for all of the organisms in the public sequence databases. The taxonomy record can be retrieved from the NCBI Taxonomy resource: https://www.ncbi.nlm.nih.gov/taxonomy/ |\n",
    "| 7      | \"species_taxid\" | Species taxonomy ID: the NCBI taxonomy identifier for the species from which the genome assembly was derived. The species taxid will differ from the organism taxid (column 6) only when the organism was reported at a sub- species or strain level. |\n",
    "| 8      | \"organism_name\" | Organism name: the scientific name of the organism from which the sequences in the genome assembly were derived. This name is taken from the NCBI Taxonomy record for the taxid specified in column 6. Some older taxids were assigned at the strain level and for these the organism name will include the strain. Current practice is only to assign taxids at the species level; for these the organism name will be just the species, however, the strain name will be reported in the infraspecific_name field (column 9). |\n",
    "| 9      | \"infraspecific_name\" | Infraspecific name: the strain, breed, cultivar or ecotype of the organism from which the sequences in the genome assembly were derived. Data are reported in the form tag=value, e.g. strain=AF16. Strain, breed, cultivar and ecotype are not expected to be used together, however, if they are then they will be reported in a list separated by \", /\". Empty if no strain, breed, cultivar or ecotype is specified on the genomic sequence records. |\n",
    "| 10     | \"isolate\" | Isolate: the individual isolate from which the sequences in the genome assembly were derived. Empty if no isolate is specified on the genomic sequence records. |\n",
    "| 11     | \"version_status\" | Version status: the release status for the genome assembly version. Values: **latest** - the most recent of all the versions for this assembly chain; **replaced** - this version has been replaced by a newer version of the assembly in the same chain; **suppressed** - this version of the assembly has been suppressed An assembly chain is the collection of all versions for the same assembly accession. |\n",
    "| 12     | \"assembly_level\" | Assembly level: the highest level of assembly for any object in the genome assembly. Values: **Complete genome** - all chromosomes are gapless and have no runs of 10 or more ambiguous bases (Ns), there are no unplaced or unlocalized scaffolds, and all the expected chromosomes are present (i.e. the assembly is not noted as having partial genome representation). Plasmids and organelles may or may not be included in the assembly but if present then the sequences are gapless. **Chromosome** - there is sequence for one or more chromosomes. This could be a completely sequenced chromosome without gaps or a chromosome containing scaffolds or contigs with gaps between them. There may also be unplaced or unlocalized scaffolds. Scaffold - some sequence contigs have been connected across gaps to create scaffolds, but the scaffolds are all unplaced or unlocalized. **Contig** - nothing is assembled beyond the level of sequence contigs |\n",
    "| 13     | \"release_type\" | Release type: whether this version of the genome assembly is a major, minor or patch release. Values: **Major** - changes from the previous assembly version result in a significant change to the coordinate system. The first version of an assembly is always a major release. Most subsequent genome assembly updates are also major releases. **Minor** - changes from the previous assembly version are limited to the following changes, none of which result in a significant change to the coordinate system of the primary assembly-unit: - adding, removing or changing a non-nuclear assembly-unit - dropping unplaced or unlocalized scaffolds - adding up to 50 unplaced or unlocalized scaffolds which are shorter than the current scaffold-N50 value - replacing a component with a gap of the same length Patch - the only change is the addition or modification of a patch assembly-unit. See the NCBI Assembly model web page (https://www.ncbi.nlm.nih.gov/assembly/ model/#asmb_def) for definitions of assembly-units and genome patches. |\n",
    "| 14     | \"genome_rep\" | Genome representation: whether the goal for the assembly was to represent the whole genome or only part of it. Values: **Full** - the data used to generate the assembly was obtained from the whole genome, as in Whole Genome Shotgun (WGS) assemblies for example. There may still be gaps in the assembly. **Partial** - the data used to generate the assembly came from only part of the genome. Most assemblies have full genome representation with a minority being partial genome representation. See the Assembly help web page (https://www.ncbi.nlm.nih.gov/assembly/help/) for reasons that the genome representation would be set to partial. |\n",
    "| 15     | \"seq_rel_date\" | Sequence release date: the date the sequences in the genome assembly were released in the International Nucleotide Sequence Database Collaboration (INSDC) databases, i.e. DDBJ, ENA or GenBank. |\n",
    "| 16     | \"asm_name\" | Assembly name: the submitter's name for the genome assembly, when one was provided, otherwise a default name, in the form ASM#####v#, is provided by NCBI. Assembly names are not unique. |\n",
    "| 17     | \"submitter\" | Submitter: the submitting consortium or first position if a list of organizations. The full submitter information is available in the NCBI BioProject resource: www.ncbi.nlm.nih.gov/bioproject/ |\n",
    "| 18     | \"gbrs_paired_asm\" | GenBank/RefSeq paired assembly: the accession.version of the GenBank assembly that is paired to the given RefSeq assembly, or vice-versa. \"na\" is reported if the assembly is unpaired. |\n",
    "| 19     | \"paired_asm_comp\" | Paired assembly comparison: whether the paired GenBank & RefSeq assemblies are identical or different. Values: **identical** - GenBank and RefSeq assemblies are identical; **different** - GenBank and RefSeq assemblies are not identical; **na** - not applicable since the assembly is unpaired |\n",
    "| 20     | \"ftp_path\" | FTP path: the path to the directory on the NCBI genomes FTP site from which data for this genome assembly can be downloaded. |\n",
    "| 21     | \"excluded_from_refseq\" | Excluded from RefSeq: reasons the assembly was excluded from the NCBI Reference Sequence (RefSeq) project, including any assembly anomalies. See: https://www.ncbi.nlm.nih.gov/assembly/help/anomnotrefseq/ |\n",
    "| 22     | \"relation_to_type_material\" | Relation to type material: contains a value if the sequences in the genome assembly were derived from type material. Values: **assembly from type material** - the sequences in the genome assembly were derived from type material; **assembly from synonym type material** - the sequences in the genome assembly were derived from synonym type material; **assembly from pathotype material** - the sequences in the genome assembly were derived from pathovar material;** **assembly designated as neotype** - the sequences in the genome assembly were derived from neotype material; **assembly designated as reftype** - the sequences in the genome assembly were derived from reference material where type material never was available and is not likely to ever be available ICTV species exemplar - the International Committee on Taxonomy of Viruses (ICTV) designated the genome assembly as the exemplar for the virus species ICTV additional isolate - the International Committee on Taxonomy of Viruses (ICTV) designated the genome assembly an additional isolate for the virus species |\n",
    "| 23     | \"asm_not_live_date\" | Assembly no longer live date: the date the assembly transitioned from version_status latest to either replaced or suppressed. When the assembly is in status latest, \"na\" is reported. |"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c291d8b",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Attempt to map NCBI fields to known [DwC terms](https://dwc.tdwg.org/list/), including extensions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c2f1e78b",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "# keep copies of DwC matched columns (just in case)\n",
    "df['ncbi_assembly_accession'] = df['# assembly_accession'] \n",
    "df['ncbi_taxid'] = df['taxid'] \n",
    "df['ncbi_species_taxid'] = df['species_taxid'] \n",
    "df['ncbi_organism_name'] = df['organism_name']\n",
    "df['ncbi_infraspecific_name'] = df['infraspecific_name']\n",
    "df['ncbi_isolate'] = df['isolate']\n",
    "df['ncbi_seq_rel_date'] = df['seq_rel_date'] \n",
    "df['ncbi_asm_name'] = df['asm_name']\n",
    "df['ncbi_submitter'] = df['submitter']\n",
    "df['ncbi_ftp_path'] = df['ftp_path']\n",
    "# create an `occurrenceID` column\n",
    "\n",
    "# 1:1 mapping to best-guess DwC and extension terms\n",
    "new_names = {\n",
    "  '# assembly_accession': 'occurrenceID',\n",
    "  'bioproject': 'ncbi_bioproject',\n",
    "  'biosample': 'ncbi_biosample',\n",
    "  'wgs_master': 'ncbi_nuccore',\n",
    "  'refseq_category': 'ncbi_refseq_category',\n",
    "  'taxid': 'taxonID',\n",
    "  'species_taxid': 'specific_host',\n",
    "  'organism_name': 'scientificName',\n",
    "  'infraspecific_name': 'cloneStrain',\n",
    "  'isolate': 'source_mat_id',\n",
    "  'version_status': 'ncbi_version_status',\n",
    "  'assembly_level': 'ncbi_assembly_level',\n",
    "  'release_type': 'ncbi_release_type',\n",
    "  'genome_rep': 'ncbi_genome_rep',\n",
    "  'seq_rel_date': 'eventDate',\n",
    "  'asm_name': 'otherCatalogNumbers',\n",
    "  'submitter': 'recordedBy',\n",
    "  'gbrs_paired_asm': 'ncbi_gbrs_paired_asm',\n",
    "  'paired_asm_comp': 'ncbi_paired_asm_comp',\n",
    "  'ftp_path': 'geneticAccessionURI',\n",
    "  'excluded_from_refseq': 'ncbi_excluded_from_refseq',\n",
    "  'relation_to_type_material': 'ncbi_relation_to_type_material',\n",
    "  'asm_not_live_date': 'ncbi_asm_not_live_date'\n",
    "}\n",
    "\n",
    "# do rename\n",
    "df.rename(\n",
    "  columns = new_names, \n",
    "  inplace = True\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78d3a7f9",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "### Merge data from NCBI BioSample database\n",
    "\n",
    "NCBI provides an XML version of its [BioSample database](https://www.ncbi.nlm.nih.gov/books/NBK169436/) at `https://ftp.ncbi.nih.gov/biosample/biosample_set.xml.gz`. By mapping the `biosample` accession number (column 3) to the data extracted from BioSample XML file, we can extract a few more columns for the DwC file. The `parse_biosample_*.py` scripts have done this and saved the data into a SQLite3 database file (in data directory)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "94cf5bdf",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 272351 entries, 0 to 272350\n",
      "Data columns (total 33 columns):\n",
      " #   Column                          Non-Null Count   Dtype         \n",
      "---  ------                          --------------   -----         \n",
      " 0   occurrenceID                    272351 non-null  object        \n",
      " 1   ncbi_bioproject                 272299 non-null  object        \n",
      " 2   ncbi_biosample                  260835 non-null  object        \n",
      " 3   ncbi_nuccore                    227449 non-null  object        \n",
      " 4   ncbi_refseq_category            272351 non-null  object        \n",
      " 5   taxonID                         272351 non-null  object        \n",
      " 6   specific_host                   272351 non-null  object        \n",
      " 7   scientificName                  272351 non-null  object        \n",
      " 8   cloneStrain                     253148 non-null  object        \n",
      " 9   source_mat_id                   18013 non-null   object        \n",
      " 10  ncbi_version_status             272351 non-null  object        \n",
      " 11  ncbi_assembly_level             272351 non-null  object        \n",
      " 12  ncbi_release_type               272351 non-null  object        \n",
      " 13  ncbi_genome_rep                 272351 non-null  object        \n",
      " 14  eventDate                       272351 non-null  datetime64[ns]\n",
      " 15  otherCatalogNumbers             272155 non-null  object        \n",
      " 16  recordedBy                      272299 non-null  object        \n",
      " 17  ncbi_gbrs_paired_asm            272351 non-null  object        \n",
      " 18  ncbi_paired_asm_comp            272351 non-null  object        \n",
      " 19  geneticAccessionURI             272351 non-null  object        \n",
      " 20  ncbi_excluded_from_refseq       509 non-null     object        \n",
      " 21  ncbi_relation_to_type_material  27668 non-null   object        \n",
      " 22  ncbi_asm_not_live_date          272351 non-null  object        \n",
      " 23  ncbi_assembly_accession         272351 non-null  object        \n",
      " 24  ncbi_taxid                      272351 non-null  object        \n",
      " 25  ncbi_species_taxid              272351 non-null  object        \n",
      " 26  ncbi_organism_name              272351 non-null  object        \n",
      " 27  ncbi_infraspecific_name         253148 non-null  object        \n",
      " 28  ncbi_isolate                    18013 non-null   object        \n",
      " 29  ncbi_seq_rel_date               272351 non-null  datetime64[ns]\n",
      " 30  ncbi_asm_name                   272155 non-null  object        \n",
      " 31  ncbi_submitter                  272299 non-null  object        \n",
      " 32  ncbi_ftp_path                   272351 non-null  object        \n",
      "dtypes: datetime64[ns](2), object(31)\n",
      "memory usage: 68.6+ MB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "117272bb",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>occurrenceID</th>\n",
       "      <th>ncbi_bioproject</th>\n",
       "      <th>ncbi_biosample</th>\n",
       "      <th>ncbi_nuccore</th>\n",
       "      <th>ncbi_refseq_category</th>\n",
       "      <th>taxonID</th>\n",
       "      <th>specific_host</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>cloneStrain</th>\n",
       "      <th>source_mat_id</th>\n",
       "      <th>...</th>\n",
       "      <th>ncbi_assembly_accession</th>\n",
       "      <th>ncbi_taxid</th>\n",
       "      <th>ncbi_species_taxid</th>\n",
       "      <th>ncbi_organism_name</th>\n",
       "      <th>ncbi_infraspecific_name</th>\n",
       "      <th>ncbi_isolate</th>\n",
       "      <th>ncbi_seq_rel_date</th>\n",
       "      <th>ncbi_asm_name</th>\n",
       "      <th>ncbi_submitter</th>\n",
       "      <th>ncbi_ftp_path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_000001215.4</td>\n",
       "      <td>PRJNA164</td>\n",
       "      <td>SAMN02803731</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000001215.4</td>\n",
       "      <td>7227</td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2014-08-01</td>\n",
       "      <td>Release 6 plus ISO1 MT</td>\n",
       "      <td>The FlyBase Consortium/Berkeley Drosophila Gen...</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_000001405.40</td>\n",
       "      <td>PRJNA168</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000001405.40</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-02-03</td>\n",
       "      <td>GRCh38.p14</td>\n",
       "      <td>Genome Reference Consortium</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_000001635.27</td>\n",
       "      <td>PRJNA169</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>10090</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000001635.27</td>\n",
       "      <td>10090</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-06-24</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>Genome Reference Consortium</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCF_000001735.4</td>\n",
       "      <td>PRJNA116</td>\n",
       "      <td>SAMN03081427</td>\n",
       "      <td>NaN</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>3702</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>ecotype=Columbia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000001735.4</td>\n",
       "      <td>3702</td>\n",
       "      <td>3702</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>ecotype=Columbia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018-03-15</td>\n",
       "      <td>TAIR10.1</td>\n",
       "      <td>The Arabidopsis Information Resource (TAIR)</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_000001905.1</td>\n",
       "      <td>PRJNA70973</td>\n",
       "      <td>SAMN02953622</td>\n",
       "      <td>AAGU00000000.3</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>9785</td>\n",
       "      <td>9785</td>\n",
       "      <td>Loxodonta africana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ISIS603380</td>\n",
       "      <td>...</td>\n",
       "      <td>GCF_000001905.1</td>\n",
       "      <td>9785</td>\n",
       "      <td>9785</td>\n",
       "      <td>Loxodonta africana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ISIS603380</td>\n",
       "      <td>2009-07-15</td>\n",
       "      <td>Loxafr3.0</td>\n",
       "      <td>Broad Institute</td>\n",
       "      <td>https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       occurrenceID ncbi_bioproject ncbi_biosample    ncbi_nuccore  \\\n",
       "0   GCF_000001215.4        PRJNA164   SAMN02803731             NaN   \n",
       "1  GCF_000001405.40        PRJNA168            NaN             NaN   \n",
       "2  GCF_000001635.27        PRJNA169            NaN             NaN   \n",
       "3   GCF_000001735.4        PRJNA116   SAMN03081427             NaN   \n",
       "4   GCF_000001905.1      PRJNA70973   SAMN02953622  AAGU00000000.3   \n",
       "\n",
       "    ncbi_refseq_category taxonID specific_host           scientificName  \\\n",
       "0       reference genome    7227          7227  Drosophila melanogaster   \n",
       "1       reference genome    9606          9606             Homo sapiens   \n",
       "2       reference genome   10090         10090             Mus musculus   \n",
       "3       reference genome    3702          3702     Arabidopsis thaliana   \n",
       "4  representative genome    9785          9785       Loxodonta africana   \n",
       "\n",
       "        cloneStrain source_mat_id  ... ncbi_assembly_accession ncbi_taxid  \\\n",
       "0               NaN           NaN  ...         GCF_000001215.4       7227   \n",
       "1               NaN           NaN  ...        GCF_000001405.40       9606   \n",
       "2               NaN           NaN  ...        GCF_000001635.27      10090   \n",
       "3  ecotype=Columbia           NaN  ...         GCF_000001735.4       3702   \n",
       "4               NaN    ISIS603380  ...         GCF_000001905.1       9785   \n",
       "\n",
       "  ncbi_species_taxid       ncbi_organism_name ncbi_infraspecific_name  \\\n",
       "0               7227  Drosophila melanogaster                     NaN   \n",
       "1               9606             Homo sapiens                     NaN   \n",
       "2              10090             Mus musculus                     NaN   \n",
       "3               3702     Arabidopsis thaliana        ecotype=Columbia   \n",
       "4               9785       Loxodonta africana                     NaN   \n",
       "\n",
       "  ncbi_isolate ncbi_seq_rel_date           ncbi_asm_name  \\\n",
       "0          NaN        2014-08-01  Release 6 plus ISO1 MT   \n",
       "1          NaN        2022-02-03              GRCh38.p14   \n",
       "2          NaN        2020-06-24                  GRCm39   \n",
       "3          NaN        2018-03-15                TAIR10.1   \n",
       "4   ISIS603380        2009-07-15               Loxafr3.0   \n",
       "\n",
       "                                      ncbi_submitter  \\\n",
       "0  The FlyBase Consortium/Berkeley Drosophila Gen...   \n",
       "1                        Genome Reference Consortium   \n",
       "2                        Genome Reference Consortium   \n",
       "3        The Arabidopsis Information Resource (TAIR)   \n",
       "4                                    Broad Institute   \n",
       "\n",
       "                                       ncbi_ftp_path  \n",
       "0  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "1  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "2  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "3  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "4  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/0...  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#df.dropna(how='all')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9f9f0a03",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "import sqlite3 as sl\n",
    "import numpy as np\n",
    "import os.path\n",
    "\n",
    "db_file_path = \"../data/biosamples_DBL.db\"\n",
    "\n",
    "if os.path.exists(db_file_path):\n",
    "    con = sl.connect(db_file_path)\n",
    "\n",
    "    # Function to add extra columns from biosample XML file\n",
    "    def addBioSampleData(row):\n",
    "        biosample_accession = row['ncbi_biosample'] if pd.notnull(row['ncbi_biosample']) else ''\n",
    "        #print(\"biosample_accession =\", biosample_accession)\n",
    "        data = ()\n",
    "        if biosample_accession:\n",
    "            try:\n",
    "                cursor = con.cursor()\n",
    "                query = \"\"\"\n",
    "                SELECT dwc_EarliestDateCollected,\n",
    "                        dwc_MaterialSample,\n",
    "                        dwc_lifeStage,\n",
    "                        dwc_materialSampleID,\n",
    "                        dwc_otherCatalogNumbers,\n",
    "                        dwc_sex,\n",
    "                        email,\n",
    "                        lab,\n",
    "                        scientificName,\n",
    "                        attrs\n",
    "                FROM entry\n",
    "                WHERE id = ?\n",
    "                \"\"\"\n",
    "                cursor.execute(query, (biosample_accession,))\n",
    "                data = cursor.fetchone() # tuple is returned\n",
    "            except sl.Error as error:\n",
    "                print(\"Failed to read data from sqlite table\", error)\n",
    "\n",
    "        if data is not None and len(data) == 10:\n",
    "            # convert empty strings to NaN (so consistent with existing data = compresses better too)\n",
    "            conv = lambda i : i or np.nan\n",
    "            list2 = [conv(i) for i in data]\n",
    "            return tuple(list2)\n",
    "        else:\n",
    "            return  tuple([np.nan] * 10) # return just NaN values as tuple\n",
    "\n",
    "    # df2 = df.head(9999).copy(deep=True) # subset data for debugging\n",
    "\n",
    "    df[['EarliestDateCollected','MaterialSample','lifeStage','materialSampleID','catalogNumber','sex',\n",
    "        'ncbi_biosample_email','ncbi_biosample_lab','ncbi_biosample_scientificName',\n",
    "        'ncbi_biosample_attributes_json']] = df.apply(addBioSampleData, axis=1, result_type='expand')\n",
    "\n",
    "    con.close()\n",
    "else:\n",
    "    print(\"SQLite3 DB file was not found. You need to run the script 'notebooks/parse_biosample_into_DB.ipynb' to generate the DB first.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1d83a528",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 272351 entries, 0 to 272350\n",
      "Data columns (total 43 columns):\n",
      " #   Column                          Non-Null Count   Dtype         \n",
      "---  ------                          --------------   -----         \n",
      " 0   occurrenceID                    272351 non-null  object        \n",
      " 1   ncbi_bioproject                 272299 non-null  object        \n",
      " 2   ncbi_biosample                  260835 non-null  object        \n",
      " 3   ncbi_nuccore                    227449 non-null  object        \n",
      " 4   ncbi_refseq_category            272351 non-null  object        \n",
      " 5   taxonID                         272351 non-null  object        \n",
      " 6   specific_host                   272351 non-null  object        \n",
      " 7   scientificName                  272351 non-null  object        \n",
      " 8   cloneStrain                     253148 non-null  object        \n",
      " 9   source_mat_id                   18013 non-null   object        \n",
      " 10  ncbi_version_status             272351 non-null  object        \n",
      " 11  ncbi_assembly_level             272351 non-null  object        \n",
      " 12  ncbi_release_type               272351 non-null  object        \n",
      " 13  ncbi_genome_rep                 272351 non-null  object        \n",
      " 14  eventDate                       272351 non-null  datetime64[ns]\n",
      " 15  otherCatalogNumbers             272155 non-null  object        \n",
      " 16  recordedBy                      272299 non-null  object        \n",
      " 17  ncbi_gbrs_paired_asm            272351 non-null  object        \n",
      " 18  ncbi_paired_asm_comp            272351 non-null  object        \n",
      " 19  geneticAccessionURI             272351 non-null  object        \n",
      " 20  ncbi_excluded_from_refseq       509 non-null     object        \n",
      " 21  ncbi_relation_to_type_material  27668 non-null   object        \n",
      " 22  ncbi_asm_not_live_date          272351 non-null  object        \n",
      " 23  ncbi_assembly_accession         272351 non-null  object        \n",
      " 24  ncbi_taxid                      272351 non-null  object        \n",
      " 25  ncbi_species_taxid              272351 non-null  object        \n",
      " 26  ncbi_organism_name              272351 non-null  object        \n",
      " 27  ncbi_infraspecific_name         253148 non-null  object        \n",
      " 28  ncbi_isolate                    18013 non-null   object        \n",
      " 29  ncbi_seq_rel_date               272351 non-null  datetime64[ns]\n",
      " 30  ncbi_asm_name                   272155 non-null  object        \n",
      " 31  ncbi_submitter                  272299 non-null  object        \n",
      " 32  ncbi_ftp_path                   272351 non-null  object        \n",
      " 33  EarliestDateCollected           188177 non-null  object        \n",
      " 34  MaterialSample                  320 non-null     object        \n",
      " 35  lifeStage                       0 non-null       float64       \n",
      " 36  materialSampleID                18294 non-null   object        \n",
      " 37  catalogNumber                   2922 non-null    object        \n",
      " 38  sex                             615 non-null     object        \n",
      " 39  ncbi_biosample_email            172056 non-null  object        \n",
      " 40  ncbi_biosample_lab              2277 non-null    object        \n",
      " 41  ncbi_biosample_scientificName   254106 non-null  object        \n",
      " 42  ncbi_biosample_attributes_json  253941 non-null  object        \n",
      "dtypes: datetime64[ns](2), float64(1), object(40)\n",
      "memory usage: 89.3+ MB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4f3cac06",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>327</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272158</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272184</th>\n",
       "      <td>50.924</td>\n",
       "      <td>4.867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272185</th>\n",
       "      <td>50.924</td>\n",
       "      <td>4.867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272186</th>\n",
       "      <td>50.924</td>\n",
       "      <td>4.867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272190</th>\n",
       "      <td>50.924</td>\n",
       "      <td>4.867</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121095 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       decimalLatitude decimalLongitude\n",
       "37                                     \n",
       "177                                    \n",
       "297                                    \n",
       "327                                    \n",
       "358                                    \n",
       "...                ...              ...\n",
       "272158                                 \n",
       "272184          50.924            4.867\n",
       "272185          50.924            4.867\n",
       "272186          50.924            4.867\n",
       "272190          50.924            4.867\n",
       "\n",
       "[121095 rows x 2 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import json\n",
    "# extract `lat_lon` and `geo_loc_name` from JSON attr to new columns\n",
    "# `lat_lon` is usually in format like `39.269 N 123.798 W`\n",
    "\n",
    "def signedDecimalDegrees(split_arr):\n",
    "  # `39.269 N 123.798 W` => `39.269, -123.798`\n",
    "  lat_str = ''\n",
    "  lon_str = ''\n",
    "  if len(split_arr) == 4:\n",
    "    sign_1 = '-' if split_arr[1] == 'S' else ''\n",
    "    sign_2 = '-' if split_arr[3] == 'W' else ''\n",
    "    lat_str = sign_1 + split_arr[0]\n",
    "    lon_str = sign_2 + split_arr[2]\n",
    "  return [lat_str, lon_str]\n",
    "  \n",
    "def extract_json_values(row):\n",
    "  lat = np.nan\n",
    "  lon = np.nan\n",
    "  locality = np.nan\n",
    "  json_attrs = row['ncbi_biosample_attributes_json'] if pd.notnull(row['ncbi_biosample_attributes_json']) else ''\n",
    "  if json_attrs:\n",
    "    json_dict = json.loads(json_attrs)\n",
    "    if 'lat_lon' in json_dict:\n",
    "      latlonArray = json_dict['lat_lon'].split(' ')\n",
    "      [lat, lon] = signedDecimalDegrees(latlonArray)\n",
    "    if 'geo_loc_name' in json_dict:\n",
    "      locality = json_dict['geo_loc_name']\n",
    "  return [lat, lon, locality]\n",
    "      \n",
    "\n",
    "df[['decimalLatitude','decimalLongitude','locality']] = df.apply(extract_json_values, axis=1, result_type='expand')\n",
    "\n",
    "df.loc[df['decimalLatitude'].notna(),[\"decimalLatitude\",\"decimalLongitude\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a362bb1c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>decimalLatitude</th>\n",
       "      <th>decimalLongitude</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>327</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272158</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272184</th>\n",
       "      <td>50.924</td>\n",
       "      <td>4.867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272185</th>\n",
       "      <td>50.924</td>\n",
       "      <td>4.867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272186</th>\n",
       "      <td>50.924</td>\n",
       "      <td>4.867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272190</th>\n",
       "      <td>50.924</td>\n",
       "      <td>4.867</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121095 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       decimalLatitude decimalLongitude\n",
       "37                                     \n",
       "177                                    \n",
       "297                                    \n",
       "327                                    \n",
       "358                                    \n",
       "...                ...              ...\n",
       "272158                                 \n",
       "272184          50.924            4.867\n",
       "272185          50.924            4.867\n",
       "272186          50.924            4.867\n",
       "272190          50.924            4.867\n",
       "\n",
       "[121095 rows x 2 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df['decimalLatitude'].notna(),[\"decimalLatitude\",\"decimalLongitude\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f113bc21",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 272351 entries, 0 to 272350\n",
      "Data columns (total 46 columns):\n",
      " #   Column                          Non-Null Count   Dtype         \n",
      "---  ------                          --------------   -----         \n",
      " 0   occurrenceID                    272351 non-null  object        \n",
      " 1   ncbi_bioproject                 272299 non-null  object        \n",
      " 2   ncbi_biosample                  260835 non-null  object        \n",
      " 3   ncbi_nuccore                    227449 non-null  object        \n",
      " 4   ncbi_refseq_category            272351 non-null  object        \n",
      " 5   taxonID                         272351 non-null  object        \n",
      " 6   specific_host                   272351 non-null  object        \n",
      " 7   scientificName                  272351 non-null  object        \n",
      " 8   cloneStrain                     253148 non-null  object        \n",
      " 9   source_mat_id                   18013 non-null   object        \n",
      " 10  ncbi_version_status             272351 non-null  object        \n",
      " 11  ncbi_assembly_level             272351 non-null  object        \n",
      " 12  ncbi_release_type               272351 non-null  object        \n",
      " 13  ncbi_genome_rep                 272351 non-null  object        \n",
      " 14  eventDate                       272351 non-null  datetime64[ns]\n",
      " 15  otherCatalogNumbers             272155 non-null  object        \n",
      " 16  recordedBy                      272299 non-null  object        \n",
      " 17  ncbi_gbrs_paired_asm            272351 non-null  object        \n",
      " 18  ncbi_paired_asm_comp            272351 non-null  object        \n",
      " 19  geneticAccessionURI             272351 non-null  object        \n",
      " 20  ncbi_excluded_from_refseq       509 non-null     object        \n",
      " 21  ncbi_relation_to_type_material  27668 non-null   object        \n",
      " 22  ncbi_asm_not_live_date          272351 non-null  object        \n",
      " 23  ncbi_assembly_accession         272351 non-null  object        \n",
      " 24  ncbi_taxid                      272351 non-null  object        \n",
      " 25  ncbi_species_taxid              272351 non-null  object        \n",
      " 26  ncbi_organism_name              272351 non-null  object        \n",
      " 27  ncbi_infraspecific_name         253148 non-null  object        \n",
      " 28  ncbi_isolate                    18013 non-null   object        \n",
      " 29  ncbi_seq_rel_date               272351 non-null  datetime64[ns]\n",
      " 30  ncbi_asm_name                   272155 non-null  object        \n",
      " 31  ncbi_submitter                  272299 non-null  object        \n",
      " 32  ncbi_ftp_path                   272351 non-null  object        \n",
      " 33  EarliestDateCollected           188177 non-null  object        \n",
      " 34  MaterialSample                  320 non-null     object        \n",
      " 35  lifeStage                       0 non-null       float64       \n",
      " 36  materialSampleID                18294 non-null   object        \n",
      " 37  catalogNumber                   2922 non-null    object        \n",
      " 38  sex                             615 non-null     object        \n",
      " 39  ncbi_biosample_email            172056 non-null  object        \n",
      " 40  ncbi_biosample_lab              2277 non-null    object        \n",
      " 41  ncbi_biosample_scientificName   254106 non-null  object        \n",
      " 42  ncbi_biosample_attributes_json  253941 non-null  object        \n",
      " 43  decimalLatitude                 121095 non-null  object        \n",
      " 44  decimalLongitude                121095 non-null  object        \n",
      " 45  locality                        235544 non-null  object        \n",
      "dtypes: datetime64[ns](2), float64(1), object(43)\n",
      "memory usage: 95.6+ MB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f9e16da4",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# save to CSV file\n",
    "df.to_csv(f\"/data/arga-data/assembly_summary_{genome_directory}.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bc74c98",
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "#df = pd.read_csv(f\"/data/arga-data/assembly_summary_{genome_directory}.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bed4d65e",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "#df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0c3259c",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "aee8b7b246df8f9039afb4144a1f6fd8d2ca17a180786b69acc140d282b71a49"
  },
  "kernelspec": {
   "display_name": "Python 3.9.8 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
